controllers/mirrorpeersecret_controller.go

```diff
@@ -0,0 +1,176 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	multiclusterv1alpha1 "github.com/red-hat-storage/odf-multicluster-orchestrator/api/v1alpha1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// MirrorPeerSecretReconciler reconciles a MirrorPeerSecret object
+type MirrorPeerSecretReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+}
+
+type SecretLabelType string
+
+const (
+	SourceLabel        SecretLabelType = "BLUE"
+	DestinationLabel   SecretLabelType = "GREEN"
+	SecretLabelTypeKey                 = "odf.orchestrator.io/secrettype"
+)
+
+var (
+	sourceSecretPredicate = predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return IsSecretSource(e.Object)
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return IsSecretSource(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return IsSecretSource(e.ObjectOld) && IsSecretSource(e.ObjectNew)
+		},
+	}
+	destinationSecretPredicate = predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return IsSecretDestination(e.Object)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return IsSecretDestination(e.ObjectOld) && IsSecretDestination(e.ObjectNew)
+		},
+	}
+)
+
+func isObjectASecretWithProvidedLabel(obj client.Object, label, value string) bool {
+	sec, ok := obj.(*corev1.Secret)
+	if !ok {
+		return false
+	}
+	lblVal, ok := sec.Labels[label]
+	// if 'ok' (ie; if provided label key is present) AND values match,
+	// then return true
+	return ok && (lblVal == value)
+}
+
+func IsSecretSource(obj client.Object) bool {
+	return isObjectASecretWithProvidedLabel(obj, SecretLabelTypeKey, string(SourceLabel))
+}
+
+func IsSecretDestination(obj client.Object) bool {
+	return isObjectASecretWithProvidedLabel(obj, SecretLabelTypeKey, string(DestinationLabel))
+}
+
+// createInternalSecret a common function to create any type secret
+func createInternalSecret(secretNameAndNamespace types.NamespacedName,
```

Comment:
Let's keep common functions in a separate package.
Add all common utilities for Secret to secrets.go.

Reply:
Most common peer secret functions are moved. Some more are yet to be arranged. Will cleanup before removing the WIP